File: shared/src/commonMain/kotlin/pe/ecolecta/domain/UsuarioInvalidoException.kt
package pe.ecolecta.domain

sealed class UsuarioInvalidoException(mensaje: String) : Exception(mensaje) {
    data object UsernameVacio : UsuarioInvalidoException("El nombre de usuario es obligatorio.")
    data object NombresVacios : UsuarioInvalidoException("Los nombres son obligatorios.")
    data object DniVacio : UsuarioInvalidoException("El DNI es obligatorio.")
    data object SinRoles : UsuarioInvalidoException("El usuario debe tener al menos un rol.")
    data object UsernameDuplicado : UsuarioInvalidoException("Ya existe un usuario con ese username.")
    data object DniDuplicado : UsuarioInvalidoException("Ya existe un usuario con ese DNI.")
}

sealed class PinInvalidoException(mensaje: String) : Exception(mensaje) {
    data object FormatoInvalido : PinInvalidoException("El PIN debe tener exactamente 4 dígitos.")
    data object NoCoincideConfirmacion : PinInvalidoException("La confirmación del PIN no coincide.")
    data object Incorrecto : PinInvalidoException("Usuario o PIN incorrecto.")
    data class Bloqueado(val bloqueadoHastaEpochMs: Long) : PinInvalidoException("Cuenta bloqueada temporalmente por intentos fallidos.")
    data object Inactivo : PinInvalidoException("El usuario está inactivo.")
}

File: web/app/Domain/Usuarios/Exceptions/UsuarioInvalidoException.php
<?php

namespace App\Domain\Usuarios\Exceptions;

use DomainException;

final class UsuarioInvalidoException extends DomainException
{
    public static function usernameVacio(): self
    {
        return new self('El nombre de usuario es obligatorio.');
    }

    public static function nombresVacios(): self
    {
        return new self('Los nombres son obligatorios.');
    }

    public static function dniVacio(): self
    {
        return new self('El DNI es obligatorio.');
    }

    public static function sinRoles(): self
    {
        return new self('El usuario debe tener al menos un rol.');
    }

    public static function usernameDuplicado(): self
    {
        return new self('Ya existe un usuario con ese nombre de usuario.');
    }

    public static function motivoBloqueoObligatorio(): self
    {
        return new self('Debes indicar el motivo del bloqueo.');
    }

    public static function sinAdministradoresActivos(): self
    {
        return new self('Esta acción dejaría el sistema sin ningún administrador activo y sin bloquear.');
    }

    public static function noPuedeAsignarRolQueNoTiene(): self
    {
        return new self('No puedes asignar o quitar el rol de Administrador si tú mismo no lo tienes.');
    }

    public static function pinInvalido(): self
    {
        return new self('El PIN debe tener entre 4 y 8 dígitos numéricos.');
    }
}
